server/etcdserver/txn/txn.go

```diff
@@ -404,23 +375,33 @@ func executeTxn(ctx context.Context, lg *zap.Logger, txnWrite mvcc.TxnWrite, rt
 	return txns, nil
 }
 
-func checkPut(rv mvcc.ReadView, lessor lease.Lessor, req *pb.PutRequest) error {
-	if req.IgnoreValue || req.IgnoreLease {
-		// expects previous key-value, error if not exist
-		rr, err := rv.Range(context.TODO(), req.Key, nil, mvcc.RangeOptions{})
+func checkPut(rv mvcc.ReadView, lessor lease.Lessor, req *pb.PutRequest) (val []byte, prevKV *mvccpb.KeyValue, leaseID lease.LeaseID, err error) {
+	val, leaseID = req.Value, lease.LeaseID(req.Lease)
+
+	if req.IgnoreValue || req.IgnoreLease || req.PrevKv {
+		resp, err := rv.Range(context.TODO(), req.Key, nil, mvcc.RangeOptions{})
 		if err != nil {
-			return err
+			return nil, nil, 0, err
 		}
-		if rr == nil || len(rr.KVs) == 0 {
-			return errors.ErrKeyNotFound
+		if req.IgnoreValue || req.IgnoreLease {
+			if resp == nil || len(resp.KVs) == 0 {
+				// ignore_{lease,value} flag expects previous key-value pair
+				return nil, nil, 0, errors.ErrKeyNotFound
+			}
 		}
+		prevKV = &resp.KVs[0]
```

Comment:
hi @serathius 
seems this line of code doesnt work just tested it

single put for a key that didnt exist with "req.PrevKv == true"
will be panic because of nil pointer dereference 

let me know if i can fix it after you merged with main
or you prefer to change it on your PR.

Thanks!

Reply:
Done